doc/sonic-swss-common/ZMQ producer-consumer state table design.md

```diff
@@ -0,0 +1,82 @@
+- [3 References](#references)
+  * [ZMQ](#zmq)
+
+# 1 Functional Requirement
+## 1.1 ZMQ client supported operations
+ - ZmqClient will send message to ZMQ.
+ - ZmqClient can reuse by multiple ZmqProducerStateTable instance.
+ - ZmqClient sendMsg() method is thread safe async method, will return immediately, ZMQ lib support async operation.
+ - ZmqClient will retry when send not success:
+    - When ZMQ socket connection broken, send API will failed and need re-connect and send again.
+    - When ZMQ send queue is full, send API will failed and need retry later.
+    - When a signal come, ZMQ send API will failed, need retry again.
+ - ZmqClient will throw exception after retry failed.
+ - ZmqClient will throw exception when ZMQ connection break.
+## 1.2 ZMQ server supported operations
+ - ZmqServer will start a receive thread and receive message from ZMQ.
+ - ZmqServer can reuse by multiple ZmqConsumerStateTable instance.
+ - When ZmqServer receive message from ZMQ, ZmqServer will:
+    - De-serialize received message.
+    - Find ZmqMessageHandler by message content.
+    - Dispatch message to ZmqMessageHandler.
+## 1.3 ZMQ producer state table supported operations
+ - Producer table will use ZmqClient to send message.
+ - Should support following operations.
+   - Set:
+        void set(const std::string &key,
+                     const std::vector<FieldValueTuple> &values,
+                     const std::string &op = SET_COMMAND,
+                     const std::string &prefix = EMPTY_PREFIX)
+   - Delete:
+        void del(const std::string &key,
+                     const std::string &op = DEL_COMMAND,
+                     const std::string &prefix = EMPTY_PREFIX)
+   - Batch Set:
+        void set(const std::vector<KeyOpFieldsValuesTuple>& values)
+   - Batch Delete:
+        void del(const std::vector<std::string>& keys)
+## 1.3 ZMQ consumer state table supported operations
+ - Consumer implement ZmqMessageHandler interface.
+ - When consumer table receive message from ZmqServer, consumer table will:
+    - Send notification to select to handle received operation.
+    - Send notification to DB update thread for write received operation to database.
+      This is a configurable feature, could turn on/off this feature in use cases requiring less memory consumption or higher performance.
+    - After send notification, continue receive next message from ZMQ.
+
+# 2 Design
+ - Diagram:
+<img src="./zmq-diagram.png" style="zoom:100%;" />
+ - Sequence:
+<img src="./zmq-sequence.png" style="zoom:100%;" />
+ - Call ZmqProducerStateTable API.
+ - ZmqProducerStateTable will send message with ZmqClient.
+ - ZmqClient will serialize operation and send to ZMQ.
+   - Return when send success.
+   - Retry when send failed.
+   - Throw exception when retry failed.
+ - ZmqServer Side:
+   - m_mqPollThread:
+   - Receive message from ZMQ.
+   - De-serialize received message then dispatch message to ZmqConsumerStateTable.
```

Comment:
>dispatch message [](http://example.com/codeflow?start=40&length=16)

Do you explain how multiple ZmqProducerStateTable uses the same channel, and how dispatching message?

Reply:
Fixed, add more detail for how to dispatch.